### ISSUE TITLE
Overall Flops are less than the modules

### ISSUE THREAD

**Issue by DuanHuiyu** [facebookresearch/fvcore#76]
Hi,
Thank you for this super-cool code.
I am using this code to calculate the flops of my model. However, I encountered a very strange problem as below. It seems that the overall flops of the whole model is less than, even some modules.
Also, it seems that some conv layers don't calculate the flops. Why does this situation happen?
I am wondering if there is any solution to help with this so strange problem.
Thank you very much.

D:\anaconda3\envs\PyTorch_py3\lib\site-packages\fvcore\nn\jit_handles.py:139: RuntimeWarning: overflow encountered in long_scalars
  flop = batch_size * out_size * Cout_dim * Cin_dim * kernel_size
D:\anaconda3\envs\PyTorch_py3\lib\collections\__init__.py:802: RuntimeWarning: overflow encountered in long_scalars
  self[elem] += count
| module                       | #parameters or shape   | #flops    |
|:-----------------------------|:-----------------------|:----------|
| model                        | 6.789M                 | 1.09G     |
|  conv1.conv2d                |  0.379M                |  1.904G   |
|   conv1.conv2d.weight        |   (256, 1478, 1, 1)    |           |
|   conv1.conv2d.bias          |   (256,)               |           |
|  conv2.conv2d                |  0.59M                 |  1.074G   |
|   conv2.conv2d.weight        |   (256, 256, 3, 3)     |           |
|   conv2.conv2d.bias          |   (256,)               |           |
|  conv3.conv2d                |  0.59M                 |  0        |
|   conv3.conv2d.weight        |   (256, 256, 3, 3)     |           |
|   conv3.conv2d.bias          |   (256,)               |           |
|  res_module                  |  3.59M                 |  3.195M   |
|   res_module.0               |   1.197M               |   1.065M  |
|    res_module.0.conv1.conv2d |    0.59M               |    0      |
|    res_module.0.conv2.conv2d |    0.59M               |    0      |
|    res_module.0.se_layer     |    16.672K             |    1.065M |
|   res_module.1               |   1.197M               |   1.065M  |
|    res_module.1.conv1.conv2d |    0.59M               |    0      |
|    res_module.1.conv2.conv2d |    0.59M               |    0      |
|    res_module.1.se_layer     |    16.672K             |    1.065M |
|   res_module.2               |   1.197M               |   1.065M  |
|    res_module.2.conv1.conv2d |    0.59M               |    0      |
|    res_module.2.conv2.conv2d |    0.59M               |    0      |
|    res_module.2.se_layer     |    16.672K             |    1.065M |
|  deconv1.conv2d              |  1.049M                |  0        |
|   deconv1.conv2d.weight      |   (256, 256, 4, 4)     |           |
|   deconv1.conv2d.bias        |   (256,)               |           |
|  deconv2.conv2d              |  0.59M                 |  1.074G   |
|   deconv2.conv2d.weight      |   (256, 256, 3, 3)     |           |
|   deconv2.conv2d.bias        |   (256,)               |           |
|  deconv3.conv2d              |  0.771K                |  12.583M  |
|   deconv3.conv2d.weight      |   (3, 256, 1, 1)       |           |
|   deconv3.conv2d.bias        |   (3,)                 |           |

**Comment by ppwwyyxx** [comment]
Please provide code to reproduce the issue

**Comment by DuanHuiyu** [comment]
Hi,
Thank you for your kind reply. I have released my simplified code below.
I am not sure why this code get this strange results.
I used "conda install -c fvcore -c iopath -c conda-forge fvcore" to install this repository, and I am using windows to run it.
Look forward to your reply.

-----------------------------------------------------------------------------------

import torch
import torch.nn as nn
from torch.nn import init
import functools
from torch.optim import lr_scheduler

from torchvision import models

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

class ConvLayer(torch.nn.Sequential):
    def __init__(self, conv, in_channels, out_channels, kernel_size, stride, padding=None, dilation=1, norm=None, act=None):
        super(ConvLayer, self).__init__()
        # padding = padding or kernel_size // 2
        padding = padding or dilation * (kernel_size - 1) // 2
        self.add_module('conv2d', conv(in_channels, out_channels, kernel_size, stride, padding, dilation=dilation))
        if norm is not None:
            self.add_module('norm', norm(out_channels))
            # self.add_module('norm', norm(out_channels, track_running_stats=True))
        if act is not None:
            self.add_module('act', act)

class TestNet(torch.nn.Module):
    def __init__(self, in_channels, out_channels, n_feats, n_resblocks, norm=nn.BatchNorm2d, 
    se_reduction=None, res_scale=1, bottom_kernel_size=3, pyramid=False, downtwice=False, skip=False):
        super(TestNet, self).__init__()
        # Initial convolution layers
        conv = nn.Conv2d
        deconv = nn.ConvTranspose2d
        act = nn.ReLU(True)

        self.conv1 = ConvLayer(conv, in_channels, n_feats, kernel_size=bottom_kernel_size, stride=1, norm=None, act=act)
        if downtwice:
            self.conv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=2, norm=norm, act=act)
        else:
            self.conv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=1, norm=norm, act=act)
        self.conv3 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=2, norm=norm, act=act)

        # Upsampling Layers
        self.deconv1 = ConvLayer(deconv, n_feats, n_feats, kernel_size=4, stride=2, padding=1, norm=norm, act=act)

        if not pyramid:
            self.deconv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=1, norm=norm, act=act)
            self.deconv3 = ConvLayer(conv, n_feats, out_channels, kernel_size=1, stride=1, norm=None, act=act)
        else:
            self.deconv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=1, norm=norm, act=act)
            self.deconv3 = ConvLayer(conv, n_feats, out_channels, kernel_size=1, stride=1, norm=None, act=act)

        self.skip = skip
        
    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = self.conv3(x)
        # x = self.res_module(x)

        x = self.deconv1(x)
        x = self.deconv2(x)
        x = self.deconv3(x)
        
        return x


net_E = TestNet(6, 3, 256, 7, norm=None, res_scale=0.1, se_reduction=8, bottom_kernel_size=1, pyramid=False).to(device)

from fvcore.nn import FlopCountAnalysis,flop_count_table,flop_count_str
inputs = torch.randn((1,6,128,128)).to(device)
print(flop_count_table(FlopCountAnalysis(net_E, inputs)))

net_E = TestNet(1475, 3, 256, 7, norm=None, res_scale=0.1, se_reduction=8, bottom_kernel_size=1, pyramid=False).to(device)
inputs = torch.randn((1,1475,128,128)).to(device)
print(flop_count_table(FlopCountAnalysis(net_E, inputs)))

**Comment by DuanHuiyu** [comment]
Hi,
Thank you for your kind reply. I have released my simplified code below.
I am not sure why this code get this strange results.
I used "conda install -c fvcore -c iopath -c conda-forge fvcore" to install this repository, and I am using windows to run it.
Look forward to your reply.

-----------------------------------------------------------------------------------
`import torch
import torch.nn as nn
from torch.nn import init
import functools
from torch.optim import lr_scheduler

from torchvision import models

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
class ConvLayer(torch.nn.Sequential):
    def __init__(self, conv, in_channels, out_channels, kernel_size, stride, padding=None, dilation=1, norm=None, act=None):
        super(ConvLayer, self).__init__()
        # padding = padding or kernel_size // 2
        padding = padding or dilation * (kernel_size - 1) // 2
        self.add_module('conv2d', conv(in_channels, out_channels, kernel_size, stride, padding, dilation=dilation))
        if norm is not None:
            self.add_module('norm', norm(out_channels))
            # self.add_module('norm', norm(out_channels, track_running_stats=True))
        if act is not None:
            self.add_module('act', act)

class TestNet(torch.nn.Module):
    def __init__(self, in_channels, out_channels, n_feats, n_resblocks, norm=nn.BatchNorm2d, 
    se_reduction=None, res_scale=1, bottom_kernel_size=3, pyramid=False, downtwice=False, skip=False):
        super(TestNet, self).__init__()
        # Initial convolution layers
        conv = nn.Conv2d
        deconv = nn.ConvTranspose2d
        act = nn.ReLU(True)

        self.conv1 = ConvLayer(conv, in_channels, n_feats, kernel_size=bottom_kernel_size, stride=1, norm=None, act=act)
        if downtwice:
            self.conv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=2, norm=norm, act=act)
        else:
            self.conv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=1, norm=norm, act=act)
        self.conv3 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=2, norm=norm, act=act)

        # Upsampling Layers
        self.deconv1 = ConvLayer(deconv, n_feats, n_feats, kernel_size=4, stride=2, padding=1, norm=norm, act=act)

        if not pyramid:
            self.deconv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=1, norm=norm, act=act)
            self.deconv3 = ConvLayer(conv, n_feats, out_channels, kernel_size=1, stride=1, norm=None, act=act)
        else:
            self.deconv2 = ConvLayer(conv, n_feats, n_feats, kernel_size=3, stride=1, norm=norm, act=act)
            self.deconv3 = ConvLayer(conv, n_feats, out_channels, kernel_size=1, stride=1, norm=None, act=act)

        self.skip = skip
        
    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = self.conv3(x)
        # x = self.res_module(x)

        x = self.deconv1(x)
        x = self.deconv2(x)
        x = self.deconv3(x)
        
        return x


net_E = TestNet(6, 3, 256, 7, norm=None, res_scale=0.1, se_reduction=8, bottom_kernel_size=1, pyramid=False).to(device)

from fvcore.nn import FlopCountAnalysis,flop_count_table,flop_count_str
inputs = torch.randn((1,6,128,128)).to(device)
print(flop_count_table(FlopCountAnalysis(net_E, inputs)))

net_E = TestNet(1475, 3, 256, 7, norm=None, res_scale=0.1, se_reduction=8, bottom_kernel_size=1, pyramid=False).to(device)
inputs = torch.randn((1,1475,128,128)).to(device)
print(flop_count_table(FlopCountAnalysis(net_E, inputs)))`

**Comment by DuanHuiyu** [comment]
Hi,
Thank you for your kind reply. I have released my simplified code below.
I am not sure why this code get this strange results.
I used "conda install -c fvcore -c iopath -c conda-forge fvcore" to install this repository, and I am using windows to run it.
Look forward to your reply.

[test_complexity.txt](https://github.com/facebookresearch/fvcore/files/6725878/test_complexity.txt)

**Comment by ppwwyyxx** [comment]
The above code does not reproduce the issue.

It prints the following for me:
```
| module                  | #parameters or shape   | #flops   |
|:------------------------|:-----------------------|:---------|
| model                   | 2.822M                 | 38.961G  |
|  conv1.conv2d           |  1.792K                |  25.166M |
|   conv1.conv2d.weight   |   (256, 6, 1, 1)       |          |
|   conv1.conv2d.bias     |   (256,)               |          |
|  conv2.conv2d           |  0.59M                 |  9.664G  |
|   conv2.conv2d.weight   |   (256, 256, 3, 3)     |          |
|   conv2.conv2d.bias     |   (256,)               |          |
|  conv3.conv2d           |  0.59M                 |  2.416G  |
|   conv3.conv2d.weight   |   (256, 256, 3, 3)     |          |
|   conv3.conv2d.bias     |   (256,)               |          |
|  deconv1.conv2d         |  1.049M                |  17.18G  |
|   deconv1.conv2d.weight |   (256, 256, 4, 4)     |          |
|   deconv1.conv2d.bias   |   (256,)               |          |
|  deconv2.conv2d         |  0.59M                 |  9.664G  |
|   deconv2.conv2d.weight |   (256, 256, 3, 3)     |          |
|   deconv2.conv2d.bias   |   (256,)               |          |
|  deconv3.conv2d         |  0.771K                |  12.583M |
|   deconv3.conv2d.weight |   (3, 256, 1, 1)       |          |
|   deconv3.conv2d.bias   |   (3,)                 |          |
| module                  | #parameters or shape   | #flops   |
|:------------------------|:-----------------------|:---------|
| model                   | 3.198M                 | 45.122G  |
|  conv1.conv2d           |  0.378M                |  6.187G  |
|   conv1.conv2d.weight   |   (256, 1475, 1, 1)    |          |
|   conv1.conv2d.bias     |   (256,)               |          |
|  conv2.conv2d           |  0.59M                 |  9.664G  |
|   conv2.conv2d.weight   |   (256, 256, 3, 3)     |          |
|   conv2.conv2d.bias     |   (256,)               |          |
|  conv3.conv2d           |  0.59M                 |  2.416G  |
|   conv3.conv2d.weight   |   (256, 256, 3, 3)     |          |
|   conv3.conv2d.bias     |   (256,)               |          |
|  deconv1.conv2d         |  1.049M                |  17.18G  |
|   deconv1.conv2d.weight |   (256, 256, 4, 4)     |          |
|   deconv1.conv2d.bias   |   (256,)               |          |
|  deconv2.conv2d         |  0.59M                 |  9.664G  |
|   deconv2.conv2d.weight |   (256, 256, 3, 3)     |          |
|   deconv2.conv2d.bias   |   (256,)               |          |
|  deconv3.conv2d         |  0.771K                |  12.583M |
|   deconv3.conv2d.weight |   (3, 256, 1, 1)       |          |
|   deconv3.conv2d.bias   |   (3,)                 |          |
```

**Comment by DuanHuiyu** [comment]
Yes, you are right. I have tested it on my ubuntu system and it gives the same results as above.
Maybe something wrong when testing it on Windows.
Anyway, I will close this issue, and thanks a lot for your help!

**Comment by ppwwyyxx** [comment]
`res_module` was in your original issue but not in the given code. It might be the code simplification that lets the issue disappear.
